src/psyclone/psyir/transformations/inline_trans.py

```diff
@@ -0,0 +1,406 @@
+        the table at the call site.
+
+        :param table: the symbol table at the call site.
+        :type table: :py:class:`psyclone.psyir.symbols.SymbolTable`
+        :param routine_table: the symbol table of the routine being inlined.
+        :type routine_table: :py:class:`psyclone.psyir.symbols.SymbolTable`
+        :param precision_map: Lists of literals, indexed by the name of the \
+            precision symbol that they use.
+        :type precision_map: Dict[str, \
+            List[:py:class:`psyclone.psyir.nodes.Literal`]]
+
+        :raises InternalError: if an imported symbol is found that has not \
+            been updated to refer to a Container at the call site.
+
+        '''
+        dummy_args = routine_table.argument_list
+
+        for old_sym in routine_table.symbols:
+
+            if old_sym in dummy_args or isinstance(old_sym, ContainerSymbol):
+                # We've dealt with Container symbols in
+                # _inline_container_symbols() and we deal with dummy arguments
+                # in apply().
+                continue
+
+            old_name = old_sym.name
+            try:
+                table.add(old_sym)
+
+            except KeyError:
+                # We have a clash with a symbol at the call site.
+                if old_sym.is_import:
+                    # This symbol is imported from a Container so should
+                    # already have been updated so as to be imported from the
+                    # corresponding container in scope at the call site.
+                    callsite_csym = table.lookup(
+                        old_sym.interface.container_symbol.name)
+                    if old_sym.interface.container_symbol is not callsite_csym:
+                        # pylint: disable=raise-missing-from
+                        raise InternalError(
+                            f"Symbol '{old_sym.name}' imported from "
+                            f"'{callsite_csym.name}' has not been updated to "
+                            f"refer to that container at the call site.")
+                else:
+                    # A Symbol with the same name already exists so we rename
+                    # the one that we are adding.
+                    new_name = table.next_available_name(
+                        old_sym.name, other_table=routine_table)
+                    routine_table.rename_symbol(old_sym, new_name)
+                    table.add(old_sym)
+
+            # Check whether this symbol is used to specify the precision of
+            # any literals.
+            if old_name in precision_map:
+                for lit in precision_map[old_name]:
+                    # A literal is immutable so create a new one with the
+                    # updated symbol as its precision.
+                    dtype = ScalarType(lit.datatype.intrinsic, old_sym)
+                    lit.replace_with(Literal(lit.value, dtype))
+
+    def validate(self, node, options=None):
```

Comment:
Have you considered cases like the one below, where there is a remapping in the shape of a passed argument. I think this is valid Fortran and can't be inlined without also expressing the remapping.
```
subroutine main
  REAL A(100, 100)
  CALL S(A(26,2),N)
end subroutine

SUBROUTINE S(X,M)
  REAL X(*)
  DO I = 1, M
     X(I) = X(I) + M
  ENDDO
END
```

Reply:
*Sigh* no, you're quite right, thanks for pointing this out. Again, for now I think I'll just reject such cases in the `validate` method.

Reply:
It turns out PSyclone generates an error for an assumed-shape dummy argument so that's not possible at the moment. However, Fortran is quite happy with:

    program main
      real :: f1(100,100), f2(5,10)
      f1 = 1
      f2 = 1
      call kern(f1, f2, 10)
      write(*,*) f1(1:20,1)
    contains
      subroutine kern(field1, field2, n)
        integer, intent(in) :: n
        real, intent(inout) :: field1(n)
        real, intent(inout) :: field2(5,*)
        field1(:n) = 0.0
        field2(:,1) = 0.0
      end subroutine kern
    end program main

I seem to remember that @hiker discovered this recently too.

Reply:
I've added a check and test for this. While doing this I realised that named arguments are also going to be a problem so I've excluded those too. I've also excluded cases where the call argument is an array subsection as that's not correctly handled yet.